contracts/Controller.sol

```diff
@@ -322,7 +340,25 @@ contract Controller is ReentrancyGuard, Ownable {
      * @dev Only account owner or operator can withdraw long option from vault
      * @param _args WithdrawArgs structure
      */
-    // function _withdrawLong(Actions.WithdrawArgs memory _args) internal isAuthorized(_args.owner) {}
+    function _withdrawLong(Actions.WithdrawArgs memory _args)
+        internal
+        isAuthorized(msg.sender, _args.owner)
+        runPreprocessing(_args.owner, _args.vaultId)
+        returns (MarginAccount.Vault memory)
+    {
+        OtokenInterface otoken = OtokenInterface(_args.asset);
+
+        require(now <= otoken.expiryTimestamp(), "Controller: can not withdraw an expired otoken");
+
+        vaults[_args.owner][_args.vaultId]._removeLong(address(otoken), _args.amount, _args.index);
+
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        marginPool.transferToUser(address(otoken), _args.to, _args.amount);
+
+        return vaults[_args.owner][_args.vaultId];
```

Comment:
I don't think we need to return this. Same for `_depositLong `, if there're multiple actions, we will just waste the gas of updating this `vault memory`.
I think we should just let `_verifyFinalState ` function access the owner's vault directly, and that pass that to the calculator.

Reply:
Some actions have no account owner/vault id arguments, so it will require checking that, and based on action assign vault owner, vault id and call the `_verifyFinalState`... will think more about it and refactor it in another PR.

Reply:
okay I got your point.